src/Socket.cpp

```diff
@@ -34,57 +35,53 @@ void Socket::write(const std::string& buff) {
-            throw std::runtime_error("Socket::read() - read() failed");
-        }
+/*
+ * Reads buffsize bytes. Returns number of bytes read.
+ */
+void Socket::read(char* buff, const size_t buffsize) {
+    if(buffsize > MAX_BUFFER_SIZE)
+        throw new std::invalid_argument("Socket::read(): buffsize too large");
 
-        buff += res;
-        buffsize -= res;
-    }
+    while(_bufferSize < buffsize)
+        _bufferedRead();
+
+    memcpy(buff, _buffer, buffsize);
+    _bufferSize -= buffsize;
+    memmove(_buffer, _buffer + buffsize, _bufferSize);
 }
 
 /*
- * Reads a line terminated by \r\n. Returns length of a received string.
- *
- * Current implementation reads bytes one by one. It's probably not the
- * best approach, but it's simple and doesn't require a temporary buffer
- * which should be somehow shifted all the time. So until this method
- * becomes a bottleneck for someone it's good enough.
+ * Reads a line terminated by \r\n or \n. Returns length of a received string.
  */
-size_t Socket::readLine(char* buff, size_t buffsize) {
-    size_t currsize = 0;
-    // TODO: optimize, figure out how many bytes are available using ioctl
-    while(currsize < buffsize) {
-        ssize_t res = ::read(_fd, buff + currsize, 1);
-        if(res <= 0) {
-            if(errno == EINTR)
-                continue;
+size_t Socket::readLine(char* buff, const size_t buffsize) {
+    size_t end = -1;
+    do {
+        end++;
+        if(end == _bufferSize)
+            _bufferedRead();
+        if(end >= buffsize)
+            throw std::runtime_error("Socket::readLine - line too long");
+    } while(_buffer[end] != '\n');
 
-            if(res == 0) /* keep this! */
-                throw std::runtime_error("Socket::read() - client closed connection");
-
-            std::ostringstream oss;
-            oss << "Socket::readLine() - read() failed, res = " << res << ", errno = " << errno;
-            throw std::runtime_error(oss.str());
-        }
+    memcpy(buff, _buffer, end);
+    buff[end] = '\0';
+    _bufferSize -= end + 1;
+    memmove(_buffer, _buffer + end + 1, _bufferSize);
+    if(end > 0 && buff[end - 1] == '\r') {
+        end--;
+        buff[end] = '\0';
+    }
+    return end;
+}
 
-        currsize++;
+void Socket::_bufferedRead() {
+    size_t count = MAX_BUFFER_SIZE - _bufferSize;
+    ssize_t res = ::read(_fd, _buffer + _bufferSize, count);
+    if(res <= 0) {
+        if(res == 0) /* keep this! */
+            throw std::runtime_error("Socket::read() - client closed connection");
 
-        if((currsize >= 2) && (buff[currsize - 1] == '\n') &&
-           ((buff[currsize - 2] == '\r') || (buff[currsize - 2] == '\n' /* see [1] */))) {
-            /* [1] - it's for debugging purpose only, in case developer decided to use `nc` */
-            buff[currsize - 2] = '\0';
-            return currsize - 2;
-        }
+        throw std::runtime_error("Socket::readLine() - read() failed: " + std::string(strerror(errno)));
```

Comment:
You should also check for the `errno == EINTR` case before throwing an exception. See the previous Socket::read implementation.